### PR TITLE
Feat (mx): unpadding during dequantization

### PR DESCRIPTION
**Pull request by Giuseppe5**
## Reason for this PR

Groupwise quantization requires padding when the input channel shape is not divisible by groupsize.

Padding works well until it doesn't, and there are important edge cases that were not covered by the previous implementation.
 (e.g., weight only quantization where padding was required. Until now, we also had to force activation quantization because otherwise we had shape mismatch).

## Changes Made in this PR
With the current implementation, we un-pad when dequantizing, taking care of all the edge cases

Few todos:
- [x] Consolidate dequantization for groupwise tensors + inference export
- [x] Fix typing in runtime activations
- [x] Testing

## Risk Highlight

- [ ] This PR includes code from another work (please detail).
- [ ] This PR contains API-breaking changes.
- [ ] This PR depends on work in another PR (please provide links/details).
- [ ] This PR introduces new dependencies (please detail).
- [ ] There are coverage gaps not covered by tests.
- [ ] Documentation updates required in subsequent PR.

## Checklist

- [x] Code comments added to any hard-to-understand areas, if applicable.
- [x] Changes generate no new warnings.
- [x] Updated any relevant tests, if applicable.
- [x] No conflicts with destination `dev` branch.
- [ ] I reviewed my own code changes.
- [x] Initial CI/CD passing.
- [ ] 1+ reviews given, and any review issues addressed and approved.
- [x] Post-review full CI/CD passing.

### CODE REVIEW

File: src/brevitas/proxy/groupwise_float_parameter_quant.py [hunk 1]
```diff
@@ -28,6 +28,7 @@ def apply_input_view(self, x):
         return x.flatten(start_dim, start_dim + 1)
 
     def create_quant_tensor(self, qt_args: Tuple[Any]) -> GroupwiseFloatQuantTensor:
+        shape = self.tracked_parameter_list[0].shape
```
Comment (Giuseppe5) [inline src/brevitas/proxy/groupwise_float_parameter_quant.py:31]: We don't support weight quant sharing for groupwise anyway, so this is safe, but it is ugly.

File: src/brevitas/quant_tensor/groupwise_float_quant_tensor.py [hunk 2]
```diff
@@ -89,19 +91,9 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
...
-            new_scale = self.scale_.expand(curr_shape).flatten(start_dim, start_dim + 1)
-        else:
-            new_scale = self.scale_
-        if self.zero_point_.shape != ():
-            new_zp = self.zero_point_.expand(curr_shape).flatten(start_dim, start_dim + 1)
-        else:
-            new_zp = self.zero_point_
-
-        return new_value, new_scale, new_zp
+        from brevitas.utils.quant_utils import groupwise_dequant
```
Comment (Giuseppe5) [inline src/brevitas/quant_tensor/groupwise_float_quant_tensor.py:94]: Also ugly, maybe the function should live somewhere else?

File: src/brevitas/proxy/groupwise_float_runtime_quant.py [hunk 3]
```diff
@@ -45,7 +45,8 @@ def create_quant_tensor(
                 inf_values,
                 nan_values,
                 self.is_signed,
-                self.training)
+                self.training,
+                dequant_shape=x.shape)
```
Comment (nickfraser) [inline src/brevitas/proxy/groupwise_float_runtime_quant.py:49]: Do we need to add to the clause that `x is not None`? I.e., won't this fail if `x=None`?

File: src/brevitas/proxy/groupwise_float_runtime_quant.py [hunk 4]
```diff
@@ -60,5 +61,6 @@ def create_quant_tensor(
                 x.inf_values,
                 x.nan_values,
                 x.signed,
-                self.training)
+                self.training,
+                x.shape)
```
Comment (nickfraser) [inline src/brevitas/proxy/groupwise_float_runtime_quant.py:65]: Same here: won't this fail if `x=None` & `qt_args` is a `Tensor`?

File: src/brevitas/proxy/float_runtime_quant.py [hunk 5]
```diff
@@ -77,7 +77,7 @@ def create_quant_tensor(
             self,
             qt_args: Union[torch.Tensor, Tuple[Any]],
             x: Optional[FloatQuantTensor] = None) -> FloatQuantTensor:
-        if x is None:
+        if isinstance(qt_args, tuple):
             out = FloatQuantTensor(*qt_args, signed=self.is_signed, training=self.training)
         else:
             out = FloatQuantTensor(
```
Comment (nickfraser) [inline src/brevitas/proxy/float_runtime_quant.py:83]: Same here: won't this fail if `x=None` & `qt_args` is a `Tensor`?

**Review by nickfraser — CHANGES_REQUESTED** [review]
Please check guards for the optional argument `x`. I think this can crash under certain circumstances. If there are preconditions that means that this can never occur, maybe add a comment about this.

File: src/brevitas/proxy/groupwise_float_runtime_quant.py (hunk 3 above)
Comment (Giuseppe5) [reply src/brevitas/proxy/groupwise_float_runtime_quant.py:49]: X is never None anymore. File runtime_quant.py, line 195 of the new file

Comment (nickfraser) [reply src/brevitas/proxy/groupwise_float_runtime_quant.py:49]: Can we update the function signature to `create_quant_tensor` then? `x: GroupwiseFloatQuantTensor)` Or will this cause some other issue that I'm not seeing?

**Review by nickfraser — APPROVED** [review]
Approve once the function signature of `create_quant_tensor` is updated.
